### ISSUE TITLE
async feature?

### ISSUE THREAD

**Issue by explodus** [google/reflectable.dart#278]
how to invoke and wait on async methods? isn't it implemented yet?

**Comment by eernstg** [comment]
There would not be anything special about that: You would invoke a method using reflection rather than directly, and you would receive the returned result. That result would be a future `f`, and you'd await it using `await f` as usual.

**Comment by explodus** [comment]
> There would not be anything special about that: You would invoke a method using reflection rather than directly, and you would receive the returned result. That result would be a future `f`, and you'd await it using `await f` as usual.

thanks, as i wrote this lines, i saw it by myself. could i ask another question?

`
class DartReflectable extends r.Reflectable {
  const DartReflectable()
      : super(
          invokingCapability,
          newInstanceCapability,
          declarationsCapability,
          reflectedTypeCapability,
          typingCapability,
          subtypeQuantifyCapability,
          superclassQuantifyCapability,
          typeAnnotationQuantifyCapability,
        );
}

@RestApi(baseUrl: 'https://rest.somewhere.com/')
@DartReflectable()
abstract class MyClient {
  ///
  factory MyClient(Dio dio, {String baseUrl}) = _MyClient;
}
`

it's a retrofit generated class, i got all things prepared/generated, but a can't get a new instance from it, cause it has no informations about _MyClient. so i can't invoke any method and can only print informations

**Comment by eernstg** [comment]
You should be able to obtain a class mirror `cm` on `MyClient`, and then `cm.newInstance('', [someDio], {#baseUrl:'someUrl'})` should yield an instance of `_MyClient`, because the constructor in `MyClient` is a factory constructor that actually redirects the call to the constructor in `_MyClient`.

However, you can't get a mirror for `_MyClient`, because reflectable cannot generate any code that would denote that class (you can't cheat on privacy in Dart). This means that you still can't call methods on an instance of `_MyClient` reflectively. I guess the declaration of `_MyClient` is generated, so you don't have the option to give it a non-private name?

**Comment by explodus** [comment]
no it doesn't work. i tried it also with generating an extra class like 

```
@DartReflectable()
class MyClientReflectable extends _MyClient {
  MyClientReflectable(Dio dio, {String? baseUrl}) : super(dio, baseUrl: baseUrl);
}
```

to get the constructor getting called/invoked 

here are some of the error messages
```
         var reflectable = const DartReflectable();
         final myClass = reflectable.reflectType(MyClient) as ClassMirror;
         instance = reflectable.reflect(
           myClass.newInstance('MyClient', [_getDio(uri: base, withBasicAuthentication: true)]),
         );

Unhandled Exception: NoSuchCapabilityError: no capability to invoke the constructor "MeyClient"
```

```
        final client = MyClient(_getDio(uri: base, withBasicAuthentication: true), baseUrl: base);
        instance = const DartReflectable().reflect(client);

Unhandled Exception: Reflecting on un-marked type "_MeyClient"
```

Retrofit for Dio make it only a private class. I also set a new feature request there.

**Comment by eernstg** [comment]
I'm not sure where the `e` in `MeyClient` and in `_MeyClient` are coming from, but they would presumably just be undefined names, so I'll assume that it is actually `MyClient` and `_MyClient`.

The first argument to `newInstance` should be the empty string rather than `MyClient`. That argument is used to specify the latter part of the constructor name when it has two parts (so you'd pass `name` to invoke a constructor named `SomeClass.name`).

Then I would expect `myClass.newInstance('', [_getDio(uri: base, withBasicAuthentication: true)])` to work (except for typos and stuff, I don't have the code needed to actually try it out).

But it is true that you cannot do `reflectable.reflect(o)` where `o` is an instance of `_MyClient`. That's what I mentioned previously:

> However, you can't get a mirror for `_MyClient`, because reflectable cannot generate any code that would denote that class

You could create an instance of `MyClientReflectable` and use that: This is not a private name, so there shouldn't be any difficulties accessing it using a mirror.

**Comment by explodus** [comment]
sorry, the `e` was a copy&paste error only

**Comment by explodus** [comment]
and no, using `MyClientReflectable` also generates errors. I will try it with the empty constructor name.

**Comment by explodus** [comment]
```
Unhandled Exception: NoSuchCapabilityError: no capability to invoke the method "testMethod"
Receiver: Instance of 'MyClientReflectable'
```

this is the result of using an instance of `MyClientReflectable`

and using `MyClient` with empty constructor name doesn't work, then i get no instance
`Unhandled Exception: Reflecting on un-marked type "_MyClient"`

**Comment by eernstg** [comment]
> using `MyClient` with empty constructor name doesn't work, then i get no instance

`myClass.newInstance('MyClient', [_getDio(uri: base, withBasicAuthentication: true)])` should yield an object `o` such that `o.runtimeType` is `_MyClient`. As I mentioned, you can't access that object reflectively, but `Reflecting on un-marked type "_MyClient"` seems to imply that you're trying to do that.

But `MyClientReflectable` is not private, and if you can do `MyClientReflectable().testMethod(...)` then you should also be able to invoke `testMethod` reflectively (assuming, of course, that the method is covered by capabilities).

**Comment by explodus** [comment]
that's my problem, it doesn't work

```
        final client = MyClientReflectable(_getDio(uri: base, withBasicAuthentication: true), baseUrl: base);
        var test = await client.testMethod();
        test.clear();
```

this works fine. but this one

```
        var reflectable = const DartReflectable();
        instance = reflectable.reflect(client);
        List<dynamic> parameters = <dynamic>[];
        var result = instance.invoke('testMethod', parameters);
```

results in 
`Unhandled Exception: NoSuchCapabilityError: no capability to invoke the method "testMethod"`

but it works if i use it native.

and it is also the generated code

```
        r.MethodMirrorImpl(
            r'testMethod',
            4325890,
            0,
            8,
            15,
            8,
            const <int>[14],
            const <int>[],
            const prefix0.DartReflectable(),
            const <Object>[const prefix4.GET('/users')]),
```

**Comment by eernstg** [comment]
That's interesting!

Note that it isn't important whether there's a method mirror for `testMethod` (method mirrors are used to explore the declarations, looking at things like return types and such, not to call methods), so you'd be looking for something like `        r'testMethod': (dynamic instance) => instance.testMethod`.

If you don't have that then it could be because there is a missing capability, or it could be a code generation bug.

**Comment by explodus** [comment]
i could separate it

if i generate the class with retrofit, i don't get the method reflected.
if i wrote the part file by my own, it is in the reflection visible and callable

```
import 'package:dio/dio.dart';
import 'package:reflectable/reflectable.dart';
import 'package:retrofit/retrofit.dart';

import 'rest.reflectable.dart';

part 'rest.g.dart';

class DartReflectable extends Reflectable {
  const DartReflectable()
      : super(
          invokingCapability,
          newInstanceCapability,
          declarationsCapability,
        );
}

const dartReflectable = DartReflectable();

@RestApi(baseUrl: 'https://test.de/rest/')
abstract class MyClient {
  factory MyClient(Dio dio, {String baseUrl}) = _MyClient;

  @GET('/test_method')
  Future<List<int>> testMethod();
}

@dartReflectable
class MyClientReflectable extends _MyClient {
  MyClientReflectable(Dio dio, {String? baseUrl}) : super(dio, baseUrl: baseUrl);
}

void main() {
  initializeReflectable();
}
```

```
// GENERATED CODE - DO NOT MODIFY BY HAND

part of 'rest.dart';

// **************************************************************************
// RetrofitGenerator
// **************************************************************************

// ignore_for_file: unnecessary_brace_in_string_interps

class _MyClient implements MyClient {
  _MyClient(this._dio, {this.baseUrl}) {
    baseUrl ??= 'https://test.de/rest/';
  }

  final Dio _dio;

  String? baseUrl;

  @override
  Future<List<int>> testMethod() async {
    const _extra = <String, dynamic>{};
    final queryParameters = <String, dynamic>{};
    final _headers = <String, dynamic>{};
    final _data = <String, dynamic>{};
    final _result = await _dio.fetch<List<dynamic>>(_setStreamType<List<int>>(
        Options(method: 'GET', headers: _headers, extra: _extra)
            .compose(_dio.options, '/test_method',
                queryParameters: queryParameters, data: _data)
            .copyWith(baseUrl: baseUrl ?? _dio.options.baseUrl)));
    final value = _result.data!.cast<int>();
    return value;
  }

  RequestOptions _setStreamType<T>(RequestOptions requestOptions) {
    if (T != dynamic &&
        !(requestOptions.responseType == ResponseType.bytes ||
            requestOptions.responseType == ResponseType.stream)) {
      if (T == String) {
        requestOptions.responseType = ResponseType.plain;
      } else {
        requestOptions.responseType = ResponseType.json;
      }
    }
    return requestOptions;
  return [];
  }
}
```
with the generated part above the testMethod isn't available 
```
class _MyClient implements MyClient {
  _MyClient(this._dio, {this.baseUrl}) {
    baseUrl ??= 'https://test.de/rest/';
  }

  final Dio _dio;

  String? baseUrl;

  @override
  Future<List<dynamic>> testMethod() async {
    return [];
  }
}
```
with the hand written simple example it is available

could it be a problem with the build runner? that the part file isn't generated and written and the reflection is trying to get generated before?

**Comment by explodus** [comment]
```
builders:
  retrofit:
    target: ":retrofit_generator"
    import: 'package:retrofit_generator/retrofit_generator.dart'
    builder_factories: ["retrofitBuilder"]
    build_extensions: {".dart": [".g.part"]}
    auto_apply: dependents
    build_to: cache
    applies_builders: ["source_gen|combining_builder"]
    runs_before:
      - package:reflectable
```

ordering through build.yaml also doesn't help. 

it could be the `GET` above
```
  @GET('/test_method')
  Future<List<int>> testMethod();
```

so that it is a parser error?

**Comment by eernstg** [comment]
> could it be a problem with the build runner? that the part file isn't generated
> and written and the reflection is trying to get generated before?

I was thinking exactly that.. and then I'd suggest to include `applies_builders` as mentioned [here](https://pub.dev/packages/build_config#adjusting-builder-ordering).

Do you have a way to check that `package:reflectable` actually denotes the reflectable builder? Perhaps it should be `:reflectable` or `reflectable:reflectable`?

**Comment by explodus** [comment]
i try every version now, of ordering the builder calls or to see if its a parser error. only if it is handwritten it is getting correctly generated reflection information and is usable. don't have an idea anymore to get it to work.

**Comment by explodus** [comment]
already done that, there are no errors. and if i use the pure instance, it works perfectly, but reflectable doesnt generate the code to get the testMethod. So i can't invoke it through reflectable.

**Comment by eernstg** [comment]
What does 'the pure instance' mean?

**Comment by explodus** [comment]
If i use `final client = MyClientReflectable(...);' directly. Then i can call `client.testMethod();`.
But i can't get it to work through InstanceMirror or ClassMirror, to get it invoked by name.

**Comment by eernstg** [comment]
The ability to invoke a method with a public name, which is declared in a private class on an instance of a non-private subtype (`extends` or `implements`) of the private class is tested in https://github.com/google/reflectable.dart/blob/master/test_reflectable/test/private_class_test.dart.

Doesn't this cover your use case?

I still suspect that the missing reflection support for `testMethod` (and whatever else is missing) is caused by the interactions of multiple code generators. Basically, the declaration isn't there when reflectable gets to look at the program (presumably because the whole part file is missing). Do you have evidence that this isn't the reason, or perhaps that it _is_ the reason?

**Comment by explodus** [comment]
That's my guess too. The multiple generators can't work together.

I'll try one last solution, to use git symlinks to the files and generate the reflection in another dart module/lib.

**Comment by eernstg** [comment]
@jakemac53, could you take a look at [this comment](https://github.com/google/reflectable.dart/issues/278#issuecomment-1175081006), and consider whether there's something about the use of the retrofit and the reflectable code generators that looks wrong?

You shouldn't have to worry about the specifics of those two code generators, just the ordering: Retrofit should run first, and reflectable should use the output from there, but it seems likely that reflectable doesn't receive the output from retrofit because it doesn't generate certain pieces of code based on declarations that retrofit should have generated.

**Comment by jakemac53** [comment]
I think probably the best solution here would be some mechanism for the _user_ to specify a builder ordering, instead of requiring the builder authors to always do about it (which adds in the coupling).

Sometimes there is already coupling, in which case the builder authors can do it for the user, but otherwise the user should be able to specify it in some way.

**Comment by jakemac53** [comment]
See https://github.com/dart-lang/build/issues/3328 for a potential solution

**Comment by natebosch** [comment]
Would it make sense for `reflectable` to use `required_inputs: [".dart"]`?

**Comment by natebosch** [comment]
> Would that mean that reflectable simply always goes after all other code generators?

Yes. I think that might be sensible for a package like `reflectable`. We do this for builders that compile the code, and reflectable is pretty closely related to compiling.

> [Edit] Just noticed that `required_inputs: [".dart"]` cannot be used by a code generator that generates any files named `*.dart` (because of the cycle: it claims to need its own output). So that won't work.

I had thought the same, but when I tried it I didn't get an error. We allow self-cycles for dependencies, we only have a restriction that a cycle can't have multiple builders. https://github.com/dart-lang/build/blob/bbcedf080a20ae0e207806c11e153d1d2be4311f/build_runner/lib/src/build_script_generate/builder_ordering.dart#L28

**Comment by eernstg** [comment]
Cf. https://github.com/google/reflectable.dart/pull/281.

**Comment by explodus** [comment]
i tried it with this build.yaml

```
targets:
  $default:
    builders:
      cs2dart_gen:
        enabled: true
      retrofit_generator|retrofit:
        enabled: true
      moor_generator:
        options:
          write_from_json_string_constructor: true
          use_data_class_name_for_companions: true
          data_class_to_companions: true
          named_parameters: true
      reflectable:
        generate_for:
          - lib/reflect/rest.dart
          - lib/my/rest.dart
        options:
          formatted: true

builders:
  retrofit:
    target: ":retrofit_generator"
    import: 'package:retrofit_generator/retrofit_generator.dart'
    builder_factories: ["retrofitBuilder"]
    build_extensions: {".dart": [".g.part"]}
    auto_apply: dependents
    build_to: source
    applies_builders: ["source_gen|combining_builder"]
    runs_before:
      - package:reflectable
  cs2dart_gen:
    target: ":cs2dart_gen"
    import: "package:cs2dart_gen/cs2dart_gen.dart"
    builder_factories: ["csFileBuilder"]
    build_extensions: {".dart": [".g.part"]}
    auto_apply: dependents
    build_to: source
    applies_builders: ["source_gen|combining_builder"]
    runs_before:
      - package:reflectable
      - package:retrofit_generator
  combining_builder:
    import: "package:source_gen/builder.dart"
    builder_factories: ["combiningBuilder"]
    build_extensions: {".dart": [".g.dart"]}
    auto_apply: none
    build_to: source
    required_inputs: [".g.part"]
    applies_builders: ["source_gen|part_cleanup"]
    runs_before:
      - package:reflectable
```

also doesnt look into generated `rest.g.dart` file

**Comment by jakemac53** [comment]
@explodus can you try out this work in progress feature to resolve this https://github.com/dart-lang/build/pull/3334?

**Comment by eernstg** [comment]
Considering 'required_inputs' in https://github.com/google/reflectable.dart/pull/281.

**Comment by jakemac53** [comment]
The new feature in package:build is now published, @explodus feel free to give that a go (will need to update to `build_runner` version `2.2.0`).

**Comment by explodus** [comment]
> The new feature in package:build is now published, @explodus feel free to give that a go (will need to update to `build_runner` version `2.2.0`).

it's working now! 

i can't instance it through myclient, caused by the private constructor of _myclient, but i can invoke it, if i use the real instance and use the reflection on it. that's everything i need. so i can test every function now und see the result as json. 

thanks for helping me out

**Comment by eernstg** [comment]
That's great, and thanks for all your input! Then I'll abandon #281.
